client/js/jsonedit.js

```diff
@@ -980,7 +980,10 @@ function jeApplyDelta(delta) {
     }
   }
 
-  jeDisplayTree();
+  if(widgetWasAdded || Object.values(delta.s).filter(s=>s===null).length)
+    jeDisplayTree();
+  else
+    jeUpdateTree(delta.s);
```

Comment:
Looks like you're right. Adding a widget is not inside the `batchStart`-`batchEnd` block in `addWidgetLocal`. I don't feel like changing that though because there's probably a reason for it.

Do you think it's slow enough to warrant some hacks?

Reply:
Yes. Adding a full deck took over 5 seconds on my machine.

Reply:
Can't `jeUpdateTree` simply update the children of the old and new parents (if they exist)? The only difficult case is when a widget is de-parented, so the whole tree gets rebuilt.

Reply:
That `batchStart` and `batchEnd` in `addWidgetLocal` were added as part of the globalUpdateRoutine changes.

Another approach here is to maintain the widget tree as a JS data structure and have the delta processing call a routine that turns the data structure into the DOM. But if you can put `batchStart` and `batchEnd` around the deck add that is probably simpler, and good enough.

Reply:
Adding a deck is still not instantaneous, but it's less than 100ms or so.